### DISCUSSION TITLE
Localized Keyboard support

### DISCUSSION THREAD

**Discussion by DreamPods** [sipper69/Control3#9] [Q&A]
First of all let me say this : great piece of kit!

I used it on multiple occasions as a crashcart for DC use but I'm facing a problem with some characters not showing up as expected.

I have a laptop with a non US EN keyboard layout. When I use Control3 I know I should revert to US EN layout and type like that but no matter what I choose as keyboard layout in windows I can not get for example the letter 'm' input on the remote machine.

Any idea what it would take to add other keyboard layout support?

Regards!

**Comment by sipper69** [comment]
Hi!

That is an interesting one.
Which keyboard layout do you use? 

Let me check. 
It's "just" fetching keyboard scan codes and make the correct mapping.

Do you use the Microsoft Store version or this repository? In case of the latter you can just make your own keyboard map.

**Reply by DreamPods** [reply]
Hi thx for the quick response.

I'm using the Microsoft Store version. Along with a laptop with a Belgian keyboard (see https://commons.wikimedia.org/wiki/File:Belgian_keyboard_layout.png)

If only I was able to do it myself I would have tried, but I'm a network and systems engineer with limited coding skills.

Regards

**Reply by sipper69** [reply]
Belgium keyboard on its way.
I'm curious
Btw. If you connect a US keyboard, does it work okay then?

**Reply by DreamPods** [reply]
Haven't found a US keyboard around so not able to test. Sorry not a lot of US keyboards around here.

**Reply by sipper69** [reply]
I have the Belgium keyboard. I'll test coming weekend. But I don't think I can find the time to work on an update though. Let's see.

**Reply by sipper69** [reply]
I used the AZERTY keyboard and tested with Windows installed on both host and remote. I played around with the layout settings.

1 (worked):  English (US) on host / French (Belgium) on remote

2 (wrong): French (Belgium) on host / English (US) on remote

3 (wrong): French (Belgium) on host / French (Belgium) on remote

The keys of both the US and French keyboard give the same scan codes. The keyboards are actually the same, only the keys have different characters/symbols printed on them.

It appears that the standard scan codes are untouched by default (US layout). When a different keyboard layout is set, Windows amends the scan codes low level.

In case of scenario 1 the remote receives untouched scan codes which are amended based on the French layout.

Scenario 2 amends the codes on the host but the remote expects default US codes.

Scenario 3  amends codes on the host which are amended again on the remote (complete mix-up).

Two solutions here.

1. The remote *has localization* (aka different layout supported) and I just switch the host to standard US when working on the remote (and back again when working on the host) 

2. The remote *is US default* and I translate the French layout scan codes back to default US. This scenario will not support the special characters, but always works on remote devices when there is no localization (possible).

Actually I could make a flag to support both scenarios.  Something like "AZERTY to ASCII".
With the tooltip "Select this option if you have an AZERTY keyboard which is not supported by the remote".

Let me see when I have some time for that. Maybe the first solution is easy, I'll add the second one and flag later. Let me see...

**Reply by DreamPods** [reply]
I don't have time to test your scenario's on my end today but will do tomorrow and report back! Thx!

But I think the host I tested with has the keyboard set to French(Belgium) which would explain why nothing worked here.

**Reply by sipper69** [reply]
Cool. 

Yes, if you add standard US English as layout on your host you can switch and test. 

But you must have localization to French on your remote as well. 

I added switching on the host, works beautifully.

If you don't have localization on the remote you should use the "AZERTY to ASCII" option. Let me see if I can code that.

**Reply by sipper69** [reply]
Added support in the latest update in the MS Appstore version. Please let me know if this solved your issue.
